database-commons/src/main/java/co/cask/CommonSchemaReader.java

```diff
@@ -0,0 +1,168 @@
+        }
+        Schema resultsetFieldSchema = resultsetField.getSchema().isNullable() ?
+          resultsetField.getSchema().getNonNullable() : resultsetField.getSchema();
+        Schema simpleSchema = field.getSchema().isNullable() ? field.getSchema().getNonNullable() : field.getSchema();
+
+        if (!resultsetFieldSchema.equals(simpleSchema)) {
+          throw new IllegalArgumentException(String.format("Schema field %s has type %s but in input record found " +
+                                                             "type %s ",
+                                                           field.getName(), simpleSchema.getType(),
+                                                           resultsetFieldSchema.getType()));
+        }
+      }
+      return schema.getFields();
+
+    }
+    return resultsetSchema.getFields();
+  }
+
+  @Override
+  public List<Schema.Field> getSchemaFields(ResultSet resultSet) throws SQLException {
+    List<Schema.Field> schemaFields = Lists.newArrayList();
+    ResultSetMetaData metadata = resultSet.getMetaData();
+    // ResultSetMetadata columns are numbered starting with 1
+    for (int i = 1; i <= metadata.getColumnCount(); i++) {
+      String columnName = metadata.getColumnName(i);
+      Schema columnSchema = getSchema(metadata, i);
+      if (ResultSetMetaData.columnNullable == metadata.isNullable(i)) {
+        columnSchema = Schema.nullableOf(columnSchema);
+      }
+      Schema.Field field = Schema.Field.of(columnName, columnSchema);
+      schemaFields.add(field);
+    }
+    return schemaFields;
+  }
+
+  @Override
+  public Schema getSchema(ResultSetMetaData metadata, int index) throws SQLException {
+    // Type.STRING covers sql types - VARCHAR,CHAR,CLOB,LONGNVARCHAR,LONGVARCHAR,NCHAR,NCLOB,NVARCHAR
+
+    int sqlType = metadata.getColumnType(index);
+
+    Schema.Type type = Schema.Type.STRING;
+    switch (sqlType) {
+      case Types.NULL:
+        type = Schema.Type.NULL;
+        break;
+
+      case Types.ROWID:
+        break;
+
+      case Types.BOOLEAN:
+      case Types.BIT:
+        type = Schema.Type.BOOLEAN;
+        break;
+
+      case Types.TINYINT:
+      case Types.SMALLINT:
+      case Types.INTEGER:
+        type = Schema.Type.INT;
+        break;
+
+      case Types.BIGINT:
+        type = Schema.Type.LONG;
+        break;
+
+      case Types.REAL:
+      case Types.FLOAT:
+        type = Schema.Type.FLOAT;
+        break;
+
+      case Types.NUMERIC:
+      case Types.DECIMAL:
+        int precision = metadata.getPrecision(index); // total number of digits
+        int scale = metadata.getScale(index); // digits after the decimal point
+        // if there are no digits after the point, use integer types
+        type = scale != 0 ? Schema.Type.DOUBLE :
+          // with 10 digits we can represent 2^32 and LONG is required
+          precision > 9 ? Schema.Type.LONG : Schema.Type.INT;
```

Comment:
Is there a reason we do this? It seems it would just be better if we were consistent and used Longs?

Reply:
This logic is from original hydrator database plugin.

Reply:
hmm ok it's probably fine to leave as is then